### PR TITLE
🤖 AUTO TTS 

### PR DESCRIPTION
**Pull request by loganhart420**
This is just something I made on my personal fork for really quick testing of different models for my datasets and thought this same style would be cool for a sort of recipe api. All the model configs are either based off pre trained models or current recipes, haven't really messed around with them and I have only tested models for one epoch as my gpu is training something else. also added a data loader tool that will load the dataset with the proper audio configs. That is also just based off pre trained model configs. It's pretty easy to add more recipes. Let me know what you think and I can work on adding more recipes, also planning on making a vocoder trainer and a speaker encoder trainer.

### CODE REVIEW

File: TTS/recipe_api/complete_recipes.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+from TTS.recipe_api.model_hub import Models
+from TTS.recipe_api.utils import data_loader
+from TTS.trainer import Trainer, TrainingArgs, init_training
+
+
+class TtsTrainer:
+    """This is trainer for calling complete recipes based off public datasets.
+    all configs are based off pretrained model configs or the model papers.
+
+    usage:
```
Comment (erogol) [inline TTS/recipe_api/complete_recipes.py:10]: This part needs to be called `Examples:`

**Comment by erogol** [comment]
This is quite impressive 🚀 

Do you think we can also create a console end-point so people run these pre-defines recipes on the terminal? We can even call it AutoTTS :)

**Review by erogol — CHANGES_REQUESTED** [review]

**Comment by loganhart420** [comment]
> This is quite impressive rocket
> 
> Do you think we can also create a console end-point so people run these pre-defines recipes on the terminal? We can even call it AutoTTS :)

AutoTTS sounds wayyyyyyyy cooler than recipe api! I'm currently working on a python script to let you just define everything from the dataset you want to train on to the model you want to use as command line args, I'm just debugging the last bit it of it currently. I'm also adding a notebook for people who use google colab.

**Comment by loganhart420** [comment]
I'm pretty content with this being the first version of the auto tts for now. Obviously as this repo grows and more models get implemented I'll continue to add but right now I want to focus on trying to make a script to export models to onnx and get them to run on onnxruntime and tensorrt(I need to do this for my own project so might as well just make a script everyone can use), and try to work on batch inference, so idk when ill add more to this, let me know if anything needs changed before it can pushed to main

**Comment by erogol** [comment]
@loganhart420 I guess you are up for the review right. Do you plan any other change?

**Comment by loganhart420** [comment]
as of right now, no, I’m working on other features before I change anything to this.

**Comment by erogol** [comment]
just a heads up,I am going to start reviewing the PR next week hopefully after solving a bunch of bugs.

File: TTS/auto_tts/complete_recipes.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        self.learning_rate = learning_rate
+        self.epochs = epochs
+        self.model = TtsModels(
+            batch_size=self.batch_size,
+            mixed_precision=self.mixed_precision,
+            learning_rate=self.learning_rate,
+            epochs=self.epochs,
+        )
+
+    def ljspeechAutoTts(
```
Comment (erogol) [inline TTS/auto_tts/complete_recipes.py:33]: Better to use `_` notation instead of Camel notation to comply with the rest of the code base.

File: TTS/auto_tts/complete_recipes.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
+from TTS.auto_tts.model_hub import TtsModels, VocoderModels
+from TTS.auto_tts.utils import data_loader
+from TTS.trainer import Trainer, TrainingArgs, init_training
+
+
+class TtsExamples:
+    """This is trainer for calling complete recipes based off public datasets.
+    all configs are based off pretrained model configs or the model papers.
+
+    usage:
```
Comment (erogol) [inline TTS/auto_tts/complete_recipes.py:10]: `Examples:` instead of `usage:` 

It'd be nice to add `Args:`  and type annotations in docstrings too

File: TTS/auto_tts/complete_recipes.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        dataset, audio = data_loader(name="vctk", path=self.data_path)
+        if model_name == "glow tts":
+            model_config = self.model.ScGlowTts(audio, dataset, speaker_file, encoder=glowtts_encoder)
+        args, config, output_path, _, c_logger, tb_logger = init_training(TrainingArgs(), model_config)
+        trainer = Trainer(args, config, output_path, c_logger, tb_logger)
+        return trainer
+
+    def SamAccentureAutoTts(self, model_name, tacotron2_model_type, forward_attention=False, location_attention=True):
+        """Tacotron2 recipes for the sam dataset, based off the pre trained model."""
+        if model_name == "tacotrn2":
```
Comment (erogol) [inline TTS/auto_tts/complete_recipes.py:89]: tacotron2

File: TTS/auto_tts/complete_recipes.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            dataset, audio = data_loader(name="ljspeech", path=self.data_path)
+            model_config = self.model.ljspeech_speedy_speech(audio, dataset)
+        elif model_name == "vits tts":
+            dataset, audio = data_loader(name="ljspeech", path=self.data_path)
+            model_config = self.model.ljspeech_vits_tts(audio, dataset)
+        args, config, output_path, _, c_logger, tb_logger = init_training(TrainingArgs(), model_config)
+        trainer = Trainer(args, config, output_path, c_logger, tb_logger)
+        return trainer
+
+    def vctkAutoTts(self, model_name, speaker_file, glowtts_encoder):
```
Comment (erogol) [inline TTS/auto_tts/complete_recipes.py:77]: Maybe rather than defining different functions for each dataset you can make the dataset an argument to the function. AFAIS, the only difference between functions is the choice of datasets.

File: TTS/auto_tts/example.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
+from TTS.auto_tts.complete_recipes import Examples
```
Comment (erogol) [inline TTS/auto_tts/example.py:1]: Maybe call this `AutoTrainer` or just `Trainer`

File: TTS/auto_tts/model_hub.py [hunk 7]
```diff
@@ -0,0 +1,543 @@
...
+        self.batch_size = batch_size
+        self.output_path = output_path
+        self.mixed_precision = mixed_precision
+        self.learning_rate = learning_rate
+        self.epochs = epochs
+
+    def single_speaker_tacotron2_base(
+        self, audio, dataset, dla=0.25, pla=0.25, ga=5.0, forward_attn=True, location_attn=True
+    ):
+        config = Tacotron2Config(
```
Comment (erogol) [inline TTS/auto_tts/model_hub.py:39]: How about fetching these configs from the real `recipes`  when exist to reduce the decoupling

File: TTS/auto_tts/model_hub.py [hunk 8]
```diff
@@ -0,0 +1,543 @@
...
+            use_noise_augment=True,
+            compute_input_seq_cache=True,
+            output_path=self.output_path,
+            datasets=[dataset],
+        )
+        return config
+
+    def ljspeech_speedy_speech(self, audio, dataset):
+        """Base speedy speech model for ljpseech dataset."""
+        model_args = SpeedySpeechArgs(
```
Comment (erogol) [inline TTS/auto_tts/model_hub.py:252]: SpeedySpeech is tricky to train since it needs character durations precomputed. You either compute them externally or train a Tacotron model first to compute durations. Maybe Speedy Speecy training should first start with the Tacotron training and compute the durations. But also it sounds like a lot of clutter in the code.

**Comment by erogol** [comment]
It'd also be nice to write some documentation to show the intended use of the `AutoTTS`. Right now, I see 2 entry points. One is using the command line arguments to train a model. Two, using python and calling the training function for a particular recipe.  Missing something ?

**Comment by erogol** [comment]
Looks like you also need to rebase the latest dev after my annoying force push of the dev

**Comment by erogol** [comment]
BTW you need to rebase to the latest dev to pass around the CLA issue.

File: TTS/auto_tts/model_hub.py (hunk 8 above)
Comment (loganhart420) [reply TTS/auto_tts/model_hub.py:254]: Yea I saw that when testing that out, for now I'm just going to leave it out until I create a stable and clean experiment function

**Comment by loganhart420** [comment]
> It'd also be nice to write some documentation to show the intended use of the `AutoTTS`. Right now, I see 2 entry points. One is using the command line arguments to train a model. Two, using python and calling the training function for a particular recipe. Missing something ?

No you are completely right in the ways to train, I'm just a really terrible documentation writer, I'm working on setting up ways to use the current recipe configs when they are present and then I'm going to completely rewrite the documentation on usage

**Comment by loganhart420** [comment]
Just to give some explanation as to why I made the if __name__ statement in the recipe files, when I tried to import the existing 
recipe configs in those files it would excite the trainer.fit() and launch training, putting everything except the configs under the if __name__ statement let me import the configs to work with auto tts and still keep the same behavior when running the actual script.

**Comment by erogol** [comment]
I'll take a look at the changes on Monday. 

BTW all of the things you mentioned above sounds amazing 👍 . 

Please go incremental and send small PRs so that we can discuss and sync on them before you do too much work. Just to keep you away from wasting your time.

**Comment by loganhart420** [comment]
sorry for such a late reply, I was catching up on my university work but I definitely only plan on doing a little at a time. i’m currently doing the data formatters because having a generic way to load data that’s just in a specific format would make fine tuning a lot easier imo. i’ll open a new pr once i make sure at least one works.

**Comment by erogol** [comment]
Just a heads up. I want to finish the new trainer API before reviewing this PR.  It will take a bit longer. But just wanted to show that I didn't forget the PR. :) 

For the data formatter, you mean something similar to the formatters we already have or you mean to format a random dataset to a specific shape. In either ways, It may be useful to check the available formatters.

**Comment by loganhart420** [comment]
my end goal is to get something similar to the fastai datablock if you’ve ever seen it, fastai was my first exposure to deep learning and I loved the way you can easily load up data with it. if you haven’t seen it basically I just want to make something that lets you identify data and labels and be able to load it into a dataset, I also want to make it return the same items as the data formatters so no one has to rewrite the tts dataset. I still don’t have it all figured out yet so right now i’m just working on formatters that will take say a csv file with speaker Ids, text and audio paths and format it for the tts dataset(but obviously for more than just csv files). i’ve looked at the current ones and yea some of them already can do this, I’m just gonna write ones for data formats the current formatters don’t support to start. eventually though the end goal is to have a generic data formatters/loader for any audio data.

**Comment by loganhart420** [comment]
and thanks for letting me know you haven’t forgot haha:). Since it’ll be a little longer i’ll try to see if i can finish my pretrained model loader since my plan was just to make it a function in the autotrainer class.

**Comment by erogol** [comment]
I skimmed the DataBlock API. It looks like an overkill for our use-case. 

TTS datasets always have transcripts and audio files somewhere. The only difference is the folder structure.

I think if we define a specific  format for uploading transcripts and the audio files then we can dodge the heavy lift of implementing such an API. 

And then after the first version we can start improving the data-loader API with some user feedback.

What do you think ?

I'd also prefer to implement a new data-loader API for the whole project, if we want to implement it.

**Comment by loganhart420** [comment]
That sounds like a great way to go about it. right now my only additional loaders I was going to write were dataloaders for different file formats mostly because i've had to deal with different formats like json files containing everything. could also probably just make a notebook on how to format custom data to ljspeech formatter. I was just thinking of ways to make it easier on users who really aren't programmers but still want to take their data and train easily

**Comment by erogol** [comment]
To me, it is easier just to define a certain folder hierarchy (for instance LJSpeech format) than make it 🐸TTS standard. Then non-tech users only need to move files to conform to the format. 

Currently, we guide people to use the LJSpeech format in our DOCs

I think allowing multiple formats does not help the non-tech user and gets us to write more code. 

What do you think?

**Comment by erogol** [comment]
Also, I was thinking maybe we should make AutoTTS a separate repo. We can get people to use it and get feedback without intensive PR work and merging it to 🐸TTS.

**Comment by loganhart420** [comment]
I am not if you’d like me to get on it I’d be more than happy to

File: TTS/auto_tts/model_hub.py [hunk 9]
```diff
@@ -0,0 +1,530 @@
...
+    ):
+
+        self.batch_size = batch_size
+        self.output_path = output_path
+        self.mixed_precision = mixed_precision
+        self.learning_rate = learning_rate
+        self.epochs = epochs
+        self.manager = ModelManager()
+
+    def _single_speaker_from_pretrained(self, model_name):
```
Comment (erogol) [inline TTS/auto_tts/model_hub.py:38]: It'd be easier to use the ModelManager to parse the models from .models.json  so we dont need to add models manually

File: TTS/auto_tts/complete_recipes.py [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+        mixed_precision=False,
+        learning_rate=1e-3,
+        epochs=1000,
+    ):
+
+        super().__init__(batch_size, mixed_precision, learning_rate, epochs, output_path)
+        self.data_path = data_path
+        self.dataset_name = dataset
+
+    def single_speaker_autotts(  # im actually going to change this to autotts_recipes and i'm making a more generic
```
Comment (erogol) [inline TTS/auto_tts/complete_recipes.py:71]: For docstrings use this format https://numpydoc.readthedocs.io/en/latest/format.html

Comment (erogol) [reply TTS/auto_tts/complete_recipes.py:71]: Even for personal notes

File: TTS/auto_tts/complete_recipes.py [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+                        dla=0.5,
+                        ga=0.0,
+                        forward_attn=forward_attention,
+                        location_attn=location_attention,
+                    )
+        args, config, output_path, _, c_logger, tb_logger = init_training(TrainingArgs(), model_config)
+        trainer = Trainer(args, config, output_path, c_logger, tb_logger)
+        return trainer
+
+    def multi_speaker_autotts(
```
Comment (erogol) [inline TTS/auto_tts/complete_recipes.py:166]: You can add ForwardTTS model too for multi-speaker

File: TTS/auto_tts/example.py [hunk 12]
```diff
@@ -0,0 +1,12 @@
+from TTS.utils.manage import ModelManager
```
Comment (erogol) [inline TTS/auto_tts/example.py:1]: I guess you want to do more here ?

File: TTS/auto_tts/single_speaker_autotts.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
+import argparse
+
+from TTS.auto_tts.complete_recipes import TtsAutoTrainer
+
+
+def main():
+    parser = argparse.ArgumentParser()
```
Comment (erogol) [inline TTS/auto_tts/single_speaker_autotts.py:7]: For Argparsing over default values you can use Coqpit https://github.com/coqui-ai/coqpit

File: TTS/auto_tts/utils.py [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+
+def custom_data_loader(sr, audio_path):
+    dataset = BaseDatasetConfig(name="ljspeech", meta_file_train="metadata.csv", path=audio_path)
+    pass
+    # this is for loading custom dataloader, it still takes the ljspeech format but the audio configs will differ
+    # with each users data so im thinking of a way to have users define their own audio params with this
+
+
+def pick_glowtts_encoder(encoder_name: str):
```
Comment (erogol) [inline TTS/auto_tts/utils.py:76]: Better to parse it from the code gain to prevent manual editing in the future.

File: TTS/auto_tts/vocoder_autotts.py [hunk 15]
```diff
@@ -0,0 +1,47 @@
+import argparse
```
[the remark this answers is not in the document]
Comment (erogol) [reply TTS/auto_tts/vocoder_autotts.py:1]: Exactly.

File: TTS/tts/datasets/dataset_downloaders.py [hunk 16]
```diff
@@ -0,0 +1,256 @@
+import logging
```
Comment (erogol) [inline TTS/tts/datasets/dataset_downloaders.py:1]: Can you separate this to a new PR ? 

You should also create separate PRs for changes under 🐸TTS as we move AutoTTS to a new repo

**Comment by erogol** [comment]
I put some comments to your changes. Let me know if you have any questions .

File: TTS/tts/datasets/dataset_downloaders.py (hunk 16 above)
Comment (loganhart420) [reply TTS/tts/datasets/dataset_downloaders.py:1]: yea I had made this before I you guys added the downloaders so I'm gong to make a new PR just adding functions for the other datasets

File: TTS/auto_tts/example.py (hunk 12 above)
Comment (loganhart420) [reply TTS/auto_tts/example.py:1]: yea I added that to the commit by mistake. I was just seeing if I understood the model manager correctly. 
I already added a way to call pretrained models in the AutoTTS repo and I'll add that in here instead of the way I made it download models

File: TTS/auto_tts/vocoder_autotts.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (loganhart420) [reply TTS/auto_tts/vocoder_autotts.py:1]: Ok cool, I'm actually going to make this part of a class that has predefined training scenarios in it. like having a method for training several tts models and showing results for each one and having another method for training a spectrogram model and vocoder in succession. I'm going to add it to the AutoTTS repo first then add the spectrogram function here after.

**Comment by erogol** [comment]
I close this as it is going to be a separate repo

**Comment by king-dahmanus** [comment]
Hey, just wanted to come here and say that auto tts is an actual tts engine for android. It basically allows for multi lingual tts by switching between tts voices. So you might need changing it
